dtfs-central-api/src/services/state-machines/utilisation-report/event-handlers/helpers/update-fee-record-payment-join-table-payment-amounts-used-for-fee-record.ts

```diff
@@ -0,0 +1,150 @@
+      sortPaymentsByAmountAscending(paymentsSortedAscending);
+      break;
+    }
+
+    joinTableEntities.push(
+      FeeRecordPaymentJoinTableEntity.create({
+        feeRecordId,
+        paymentId: largestPayment.id,
+        paymentAmountUsedForFeeRecord: largestPayment.amount,
+      }),
+    );
+  }
+  return joinTableEntities;
+};
+
+/**
+ * Gets the updated fee record payment join table
+ * entities by greedily splitting payment amounts
+ * across the fee record fees sorted in descending
+ * order
+ * @param param - The fee record payment entity group
+ * @param param.feeRecords - The fee record entities
+ * @param param.payments - The payment entities
+ * @returns The updated entity
+ */
+const getUpdatedFeeRecordPaymentJoinTableEntitiesFromFeeRecordPaymentGroup = ({
+  feeRecords,
+  payments,
+}: FeeRecordPaymentEntityGroup): FeeRecordPaymentJoinTableEntity[] => {
+  const feeRecordsSortedDescending = orderBy(feeRecords, [(feeRecord) => feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency()], ['desc']);
+  sortPaymentsByAmountAscending(payments);
+
+  return feeRecordsSortedDescending.reduce((allJoinTableEntities, feeRecord, feeRecordIndex) => {
+    const feeRecordId = feeRecord.id;
+
+    if (feeRecordIndex === feeRecords.length - 1) {
+      const joinTableEntities = mapRemainingPaymentsToFeeRecordPaymentJoinTableEntities(payments, feeRecordId);
+      return [...allJoinTableEntities, ...joinTableEntities];
+    }
+
+    const joinTableEntities = getFeeRecordPaymentJoinTableEntitiesForFeeRecordAndPayments(feeRecord, payments);
+    return [...allJoinTableEntities, ...joinTableEntities];
+  }, [] as FeeRecordPaymentJoinTableEntity[]);
+};
+
+/**
+ * Updates the fee record payment join table payment amounts
+ * used for the corresponding fee record
+ * @param matchFeeRecords - Fee records at the MATCH status
+ * @param entityManager - The entity manager
+ */
+export const updateFeeRecordPaymentJoinTablePaymentAmountsUsedForFeeRecord = async (
+  matchFeeRecords: FeeRecordEntity[],
+  entityManager: EntityManager,
+): Promise<void> => {
+  const matchFeeRecordPaymentGroups = getFeeRecordPaymentEntityGroupsFromFeeRecordEntities(matchFeeRecords);
+  const updatedFeeRecordPaymentJoinTableEntities = matchFeeRecordPaymentGroups.reduce(
+    (acc, entityGroup) => [...acc, ...getUpdatedFeeRecordPaymentJoinTableEntitiesFromFeeRecordPaymentGroup(entityGroup)],
+    [] as FeeRecordPaymentJoinTableEntity[],
+  );
+
+  await Promise.all(
+    updatedFeeRecordPaymentJoinTableEntities.map(async ({ feeRecordId, paymentId, paymentAmountUsedForFeeRecord }) => {
+      try {
+        await entityManager.update(FeeRecordPaymentJoinTableEntity, { feeRecordId, paymentId }, { paymentAmountUsedForFeeRecord });
+      } catch (error) {
+        console.error(`Failed to update fee record payment join table entity with feeRecordId '${feeRecordId}' and paymentId '${paymentId}':`, error);
+        throw error;
+      }
+    }),
+  );
+};
```

Comment:
The reason I wanted to avoid using multiple files for this (even with the payment splitting) is because there isn't really a way to get around separating the payment splitting logic from mutating the list of payments without making code that is much harder to read than what I have currently written in my opinion (ie. passing around array indices or creating new versions of the payments array in memory which don't actually get cleared by the GC due to the objects in the array themselves not going out of scope). By keeping everything in the same file, I think it's much easier to figure out what the side effects of these functions are. If we want to highlight that splitting is taking place, I think changing the function name and return type would suffice, but I am hesistant to remove functions with side effects from the context where the side effects make sense.